### PR TITLE
Add Checkbox to "show/hide" custom fields in Gutenberg Editor

### PR DESCRIPTION
**Pull request by mitogh**
This changes requires a PR for `release/M18.15` to be merged in order for the hooks that inject the view into the settings page to be rendered, specifically this PR:

- https://github.com/moderntribe/events-pro/pull/1020

### CODE REVIEW

**Comment by mitogh** [comment]
Once https://github.com/moderntribe/events-gutenberg/pull/391 is merged we can add a change log into the next section to avoid a merge conflict or we can create another PR for the change log alone.

File: plugins/events-pro/src/Tribe/Editor.php [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+ */
+class Tribe__Gutenberg__Events_Pro__Editor extends Tribe__Gutenberg__Common__Editor {
+	
+	/**
+	 * Attach hooks into the editor
+	 *
+	 * @since TBD
+	 */
+	public function hook() {
+		add_action( 'tribe_events_pro_after_custom_field_content',
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Editor.php]: Why not a oneliner?

File: plugins/events-pro/src/Tribe/Editor.php [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+			3
+		);
+		add_filter( 'tribe-events-save-options', array( $this, 'save_custom_field_values' ) );
+		add_action( 'admin_print_styles', array( $this, 'admin_enqueue_styles' ) );
+		add_action( 'admin_print_footer_scripts', array( $this, 'admin_print_scripts' ) );
+	}
+	
+	/**
+	 * Attach a new input after each custom field input is rendered, the value is being stored in a hidden field and
+     * creates a fake <button> to have A11y benefits like focus and so on. The "fake checkbox" is used as we need to send
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Editor.php]: Odd indentation

File: plugins/events-pro/src/Tribe/Editor.php [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	 * @param $ticket
+	 * @param $index
+	 * @param $count
+	 */
+	public function after_custom_field_content( $ticket, $index, $count ) {
+		$input_id = 'gutenberg_editor_' . esc_attr( $index );
+		$value    = isset( $ticket['gutenberg_editor'] ) && $ticket['gutenberg_editor'] ? '1' : '0';
+		$class_name = $value === '1' ? 'tribe-custom-field-gutenberg-checkbox--checked' : '';
+		?>
+        <div class="tribe-field-row tribe-field-gutenberg-options">
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Editor.php]: Move this into an `admin-view` file

File: plugins/events-pro/src/Tribe/Editor.php [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+	 */
+	public function admin_enqueue_styles() {
+		/**
+		 * @todo replace with real plugin template, once is moved out of the extension
+		 */
+		$directory   = 'plugins/events-pro/src/styles/admin.css';
+		$plugins_url = plugins_url(
+			trailingslashit( basename( tribe( 'gutenberg' )->plugin_path ) )
+		);
+		wp_enqueue_style(
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Editor.php]: Use `tribe_asset`

File: plugins/events-pro/src/Tribe/Editor.php [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+	/**
+	 * Attach JS behavior into the "fake checkbox" and into the "Add Another" button in order to reset and attach behaviors
+     * into each "fake checkbox" as HTML inputs are recreated via JS on PRO side, also this allow to have a single JS logic
+     * for all the fields instead of one per field.
+	 *
+	 * @since TBD
+	 */
+	public function admin_print_scripts() {
+		?>
+        <script type="text/javascript">
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Editor.php]: Use a `tribe_asset` JS file.

File: plugins/events-pro/src/Tribe/Provider.php [hunk 6]
```diff
@@ -12,17 +12,20 @@ public function register() {
 
 		// Setup to check if gutenberg is active
 		$this->container->singleton( 'gutenberg.events-pro.plugin', 'Tribe__Gutenberg__Events_Pro__Plugin' );
-
-		if ( ! tribe( 'gutenberg.common.editor' )->should_load_blocks() ) {
+		
+		// Return if we shouldn't load blocks or Events Pro Plugin is active
+		if ( ! tribe( 'gutenberg.common.editor' )->should_load_blocks()
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Provider.php]: Break each conditional into a line.

File: plugins/events-pro/src/styles/admin.css [hunk 7]
```diff
@@ -0,0 +1,25 @@
+.tribe-custom-field-gutenberg-checkbox {
```
Comment (bordoni) [inline plugins/events-pro/src/styles/admin.css]: We need this to be generated from a `.pcss` file.

File: plugins/events-pro/src/Tribe/Editor.php [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+		foreach ( $options['custom-fields'] as $index => $field ) {
+			$checked = isset( $gutenberg_fields[ $index ] ) && '1' === $gutenberg_fields[ $index ];
+			$options['custom-fields'][ $index ]['gutenberg_editor'] = $checked;
+		}
+		return $options;
+	}
+	
+	/**
+	 * Make sure the keys of the gutenberg custom fields match the same logic as the custom fields, this logic is
+     * basically if the key or index of a gutenberg field has `_` at the start it means it belongs to an existing
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Editor.php]: Odd indentation

File: plugins/events-pro/src/Tribe/Editor.php [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+                    data-count="<?php echo esc_attr( $count ); ?>"
+                    value="<?php echo esc_attr( $value ); ?>"
+            />
+        </div>
+		<?php
+	}
+	
+	/**
+	 * Hook into the options before they are stored in the DB, and access the variable used to hold the reference
+     * to the hidden fields, from there we just need to make sure if the fields has been checked or not and update the
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Editor.php]: Odd indentation

File: plugins/events-pro/src/Tribe/Editor.php [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+				$assigned_index = ++$max_index;
+			}
+			$mapped[ $assigned_index ] = $field;
+		}
+		return $mapped;
+    }
+	
+	/**
+	 * Return the highest number for the custom fields, this value is created and updated by PRO. Fallback to a zero
+     * value if is not present on the settings or there are no custom fields present yet.
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Editor.php]: Odd indentation

File: plugins/events-pro/src/Tribe/Editor.php [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+		} else if ( isset( $current_options['custom-fields'] ) ) {
+			return count( $current_options['custom-fields'] ) + 1;
+		} else {
+			return 0;
+		}
+	}
+	
+	/**
+	 * Include styles used to style the admin elements added by the gutenberg extension on settings page or
+     * admin screens other that the ones non related to blocks
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Editor.php]: Odd indentation

File: plugins/events-pro/src/Tribe/Editor.php [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+			$plugins_url . $directory,
+			array(),
+			apply_filters( 'tribe_events_pro_css_version', Tribe__Events__Pro__Main::VERSION )
+		);
+	}
+	
+	/**
+	 * Attach JS behavior into the "fake checkbox" and into the "Add Another" button in order to reset and attach behaviors
+     * into each "fake checkbox" as HTML inputs are recreated via JS on PRO side, also this allow to have a single JS logic
+     * for all the fields instead of one per field.
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Editor.php]: Odd indentation

File: plugins/events-pro/src/Tribe/Editor.php (hunk 1 above)
Comment (mitogh) [reply plugins/events-pro/src/Tribe/Editor.php]: A single line was longer to fit into 100 chars so I split it into multiple lines instead.

**Comment by mitogh** [comment]
Thanks changes has been addressed.

File: plugins/events-pro/src/Tribe/Editor.php [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+			'value'      => $value,
+			'index'      => $index,
+			'class_name' => $value === '1' ? 'tribe-custom-field-gutenberg-checkbox--checked' : '',
+			'count'      => $count,
+		);
+		
+		$html = tribe( 'gutenberg.events.pro.admin.template' )->template( array( 'custom-fields', 'gutenberg' ), $args, false );
+		
+		tribe_asset_enqueue( 'gutenberg-events-pro-admin-additional-fields-style' );
+		tribe_asset_enqueue( 'gutenberg-events-pro-admin-additional-fields-behavior' );
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Editor.php]: Do we still need this?

File: plugins/events-pro/src/Tribe/Template/Admin.php [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+ * @since TBD
+ */
+class Tribe__Gutenberg__Events_Pro__Template__Admin extends Tribe__Template {
+	/**
+	 * Building of the Class template configuration
+	 *
+	 * @since TBD
+	 */
+	public function __construct() {
+		$this->set_template_origin( tribe( 'gutenberg' ) );
```
Comment (bordoni) [inline plugins/events-pro/src/Tribe/Template/Admin.php]: Pass `tribe( 'gutenberg.events.pro' )` to make the template folder:
`$this->set_template_folder( 'src/admin-view' );`

File: plugins/events-pro/src/Tribe/Editor.php (hunk 13 above)
Comment (mitogh) [reply plugins/events-pro/src/Tribe/Editor.php]: We shouldn't let me double check. 👍

**Pull request merged**